controllers/build.go

```diff
@@ -0,0 +1,280 @@
+func paramsForWebhookBasedBuilds(component appstudiov1alpha1.Component) []tektonapi.Param {
+	sourceCode := component.Spec.Source.GitSource.URL
+	outputImage := normalizeOutputImageURL(component.Spec.Build.ContainerImage)
+
+	params := []tektonapi.Param{
+		{
+			Name: "git-url",
+			Value: tektonapi.ArrayOrString{
+				Type:      tektonapi.ParamTypeString,
+				StringVal: sourceCode,
+			},
+		},
+		{
+			Name: "output-image",
+			Value: tektonapi.ArrayOrString{
+				Type:      tektonapi.ParamTypeString,
+				StringVal: outputImage,
+			},
+		},
+	}
+
+	return params
+}
+
+func commonAnnotations() map[string]string {
+	annotations := map[string]string{
+		"build.appstudio.openshift.io/build":   "true",
+		"build.appstudio.openshift.io/type":    "build",
+		"build.appstudio.openshift.io/version": "0.1",
+	}
+	return annotations
+}
+
+// commonStorage returns the PVC that would be created per namespace for
+// user-triggered and webhook-triggered Tekton workspaces.
+func commonStorage(name string, namespace string) corev1.PersistentVolumeClaim {
+	fsMode := corev1.PersistentVolumeFilesystem
+
+	workspaceStorage := &corev1.PersistentVolumeClaim{
+		ObjectMeta: v1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Annotations: commonAnnotations(),
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{
+				"ReadWriteOnce",
+			},
+			Resources: corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					"storage": resource.MustParse("10Mi"),
+				},
+			},
+			VolumeMode: &fsMode,
+		},
+	}
+	return *workspaceStorage
+}
+
+// route returns the Route resource that would enable
+// ingress traffic into the webhook endpoint ( aka EventListener)
+// TODO: This needs to be secure.
+func route(component appstudiov1alpha1.Component) routev1.Route {
+	var port int32 = 8080
+	webhook := &routev1.Route{
+		ObjectMeta: v1.ObjectMeta{
+			Name:        component.Name,
+			Namespace:   component.Namespace,
+			Annotations: commonAnnotations(),
+		},
+		Spec: routev1.RouteSpec{
+			Path: "/",
+			To: routev1.RouteTargetReference{
+				Kind: "Service",
+				Name: "el-" + component.Name,
+			},
+			Port: &routev1.RoutePort{
+				TargetPort: intstr.IntOrString{IntVal: port},
+			},
+		},
+	}
+	return *webhook
+}
+
+// triggerTemplate generates the TriggerTemplate resources
+// which defines how a webhook-based trigger event would be handled -
+// In this case, a PipelineRun to build an image would be created.
+func triggerTemplate(component appstudiov1alpha1.Component) (*triggersapi.TriggerTemplate, error) {
+	webhookBasedBuildTemplate := determineBuildExecution(component, paramsForWebhookBasedBuilds(component), "$(tt.params.git-revision)")
+
+	resoureTemplatePipelineRun := tektonapi.PipelineRun{
+		ObjectMeta: v1.ObjectMeta{
+			GenerateName: component.Name + "-",
+			Namespace:    component.Namespace,
```

Comment:
Just realizing something here. So since this PipelineRun will run in the namespace the component has been created in, and since components can be created in any namespace, this means each namespace will need to have the `redhat-appstudio-registry-pull-secret` secret created, right?

Probably not an issue for Milestone 2, but something to think about. @sbose78 wdyt?

Reply:
Correct. 
This is why we granted create & delete permissions to users for `secrets` in app studio namespaces.

However, it's a painful user experience because the UI doesn't have a way to take the input image (yet)- therefore the user will need to be able to push to the pre-defined image repository for her token to work. It gets easier if done from the CLI - since the user can specify both the image repo ( .spec.build.containerImage ) and create the `secret`

Given that, users don't have a way to view the contents of secrets in their namespaces, I'm thinking of having the  `redhat-appstudio-registry-pull-secret` created in every namespace at the time of the component creation ( same way the `appstudio` pvc is being setup now per namespace in this PR ). They can of course delete it and create their own secret if they wish to. What do you think ?

Mid-term/Long-term, I'd consider setting up of the namespace/workspace specific assets at the time of workspace/namespace provisioning - and not at the time of Component creation.

Reply:
👍 for both suggestions.